homeassistant/components/fritz/config_flow.py

```diff
@@ -0,0 +1,244 @@
+        """Initialize FRITZ!Box Tools flow."""
+        self._host = None
+        self._entry = None
+        self._name = None
+        self._password = None
+        self._port = None
+        self._username = None
+        self.import_schema = None
+        self.fritz_tools = None
+
+    async def fritz_tools_init(self):
+        """Initialize FRITZ!Box Tools class."""
+        self.fritz_tools = FritzBoxTools(
+            hass=self.hass,
+            host=self._host,
+            port=self._port,
+            username=self._username,
+            password=self._password,
+        )
+
+        try:
+            await self.fritz_tools.async_setup()
+        except FritzSecurityError:
+            return ERROR_AUTH_INVALID
+        except FritzConnectionException:
+            return ERROR_CONNECTION_ERROR
+
+        return None
+
+    @callback
+    def _async_create_entry(self):
+        """Async create flow handler entry."""
+        return self.async_create_entry(
+            title=self._name,
+            data={
+                CONF_HOST: self.fritz_tools.host,
+                CONF_PASSWORD: self.fritz_tools.password,
+                CONF_PORT: self.fritz_tools.port,
+                CONF_USERNAME: self.fritz_tools.username,
+            },
+        )
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a flow initialized by discovery."""
+        ssdp_location = urlparse(discovery_info[ATTR_SSDP_LOCATION])
+        self._host = ssdp_location.hostname
+        self._port = ssdp_location.port
+        self._name = discovery_info.get(ATTR_UPNP_FRIENDLY_NAME)
+        self.context[CONF_HOST] = self._host
+
+        uuid = discovery_info.get(ATTR_UPNP_UDN)
+        if uuid:
+            if uuid.startswith("uuid:"):
+                uuid = uuid[5:]
+            await self.async_set_unique_id(uuid)
+            self._abort_if_unique_id_configured({CONF_HOST: self._host})
+
+        for progress in self._async_in_progress():
+            if progress.get("context", {}).get(CONF_HOST) == self._host:
+                return self.async_abort(reason="already_in_progress")
+
+        for entry in self._async_current_entries(include_ignore=False):
+            if entry.data[CONF_HOST] == self._host:
+                if uuid and not entry.unique_id:
+                    self.hass.config_entries.async_update_entry(entry, unique_id=uuid)
+                return self.async_abort(reason="already_configured")
+
+        self.context["title_placeholders"] = {
+            "name": self._name.replace("FRITZ!Box ", "")
+        }
+        return await self.async_step_confirm()
+
+    async def async_step_confirm(self, user_input=None):
+        """Handle user-confirmation of discovered node."""
+
+        if user_input is None:
+            return self._show_setup_form_confirm()
+
+        errors = {}
+
+        self._username = user_input.get(CONF_USERNAME)
+        self._password = user_input.get(CONF_PASSWORD)
+
```

Comment:
These should also be required so `[]` instead of `.get()`

Reply:
Done

Reply:
Done